inter/block.go

```diff
@@ -1,45 +1,188 @@
+	Epoch idx.Epoch
+
+	// The hash of this block, cached on first access.
+	hash common.Hash
+}
+
+// Hash computes the hash of this block, committing all its fields.
+func (b *Block) Hash() common.Hash {
+	if b.hash == (common.Hash{}) {
+		b.hash = b.GetEthereumHeader().Hash()
+	}
+	return b.hash
+}
+
+// uncleHash is the hash to be used for the uncle field in Ethereum headers if
+// there are no uncles. See https://eips.ethereum.org/EIPS/eip-4844.
+var uncleHash = common.BytesToHash(hexutil.MustDecode(
+	"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
+))
+
+// GetEthereumHeader returns the Ethereum header corresponding to this block.
+func (b *Block) GetEthereumHeader() *types.Header {
+	return &types.Header{
+		ParentHash:  b.ParentHash,
+		UncleHash:   uncleHash,
+		Coinbase:    common.Address{}, // < in Sonic, the coinbase is always 0
+		Root:        b.StateRoot,
+		TxHash:      b.TransactionsHashRoot,
+		ReceiptHash: b.ReceiptsHashRoot,
+		Bloom:       b.LogBloom,
+		Difficulty:  big.NewInt(int64(b.Difficulty)),
+		Number:      big.NewInt(int64(b.Number)),
+		GasLimit:    b.GasLimit,
+		GasUsed:     b.GasUsed,
+		Time:        uint64(b.Time.Time().Unix()),
+		Extra:       nil, // TODO: fill in extra data required for gas computation
+		MixDigest:   b.PrevRandao,
+		Nonce:       types.BlockNonce{}, // constant 0 in Ethereum
+		BaseFee:     b.BaseFee,
+
+		// Sonic does not have a beacon chain and no withdrawals.
+		WithdrawalsHash: &types.EmptyWithdrawalsHash,
+
+		// Sonic does not support blobs, so no blob gas is used and there is
+		// no excess blob gas.
+		BlobGasUsed:   new(uint64), // = 0
+		ExcessBlobGas: new(uint64), // = 0
+	}
 }
 
 func (b *Block) EstimateSize() int {
-	return (len(b.Events)+len(b.InternalTxs)+len(b.Txs)+1+1)*32 + len(b.SkippedTxs)*4 + 8 + 8
+	return int(unsafe.Sizeof(*b)) +
+		len(b.TransactionHashes)*int(unsafe.Sizeof(common.Hash{}))
 }
 
-// GetPrevRandao computes prevrandao if not already computed and returns it.
-func (b *Block) GetPrevRandao() common.Hash {
-	return computePrevRandao(b.Events)
+// ----------------------------------------------------------------------------
+// BlockBuilder
+// ----------------------------------------------------------------------------
+
+type BlockBuilder struct {
+	block        Block
+	transactions types.Transactions
+	receipts     types.Receipts
 }
 
-func FilterSkippedTxs(txs types.Transactions, skippedTxs []uint32) types.Transactions {
-	if len(skippedTxs) == 0 {
-		// short circuit if nothing to skip
-		return txs
-	}
-	skipCount := 0
-	filteredTxs := make(types.Transactions, 0, len(txs))
-	for i, tx := range txs {
-		if skipCount < len(skippedTxs) && skippedTxs[skipCount] == uint32(i) {
-			skipCount++
-		} else {
-			filteredTxs = append(filteredTxs, tx)
-		}
+func NewBlockBuilder() *BlockBuilder {
+	return &BlockBuilder{}
+}
+
+func (b *BlockBuilder) SetNumber(number uint64) *BlockBuilder {
```

Comment:
nit: builder pattern usually writes "with" instead of "set"

Reply:
updated